x/pairing/keeper/scores/score.go

```diff
@@ -0,0 +1,222 @@
+}
+
+// TODO: currently we'll use weight=1 for all reqs. In the future, we'll get it from policy
+func GetStrategy() ScoreStrategy {
+	return uniformStrategy
+}
+
+// CalcPairingScore calculates the final pairing score for a pairing slot (with strategy)
+// For efficiency purposes, we calculate the score on a diff slot which represents the diff reqs of the current slot
+// and the previous slot
+func CalcPairingScore(scores []*PairingScore, strategy ScoreStrategy, diffSlot *PairingSlot, minStake sdk.Int) error {
+	// calculate the score for each req for each provider
+	for _, req := range diffSlot.Reqs {
+		reqName := req.GetName()
+		weight := strategy[reqName] // not checking if key is found because it's verified in init()
+
+		for _, score := range scores {
+			newScoreComp := req.Score(*score.Provider)
+			if newScoreComp == 0 {
+				err := fmt.Errorf("score component is zero. score component name: %s, provider address: %s", reqName, score.Provider.Address)
+				panic(err)
+			}
+			newScoreComp = commontypes.SafePow(newScoreComp, weight)
+
+			// divide by previous score component (if exists) and multiply by new score
+			prevReqScoreComp, ok := score.ScoreComponents[reqName]
+			if ok {
+				score.Score /= prevReqScoreComp
+			}
+			score.Score *= newScoreComp
+
+			// update the score component map
+			score.ScoreComponents[reqName] = newScoreComp
+		}
+	}
+
+	return nil
+}
+
+// PrepareHashData prepares the hash needed in the pseudo-random choice of providers
+func PrepareHashData(projectIndex string, chainID string, epochHash []byte) []byte {
+	hashData := []byte{}
+
+	// add the session start block hash to the function to make it as unpredictable as we can
+	hashData = append(hashData, epochHash...)
+	hashData = append(hashData, chainID...)      // to make this pairing unique per chainID
+	hashData = append(hashData, projectIndex...) // to make this pairing unique per consumer
+
+	return hashData
+}
+
+// PickProviders pick a <group-count> providers set with a pseudo-random weighted choice (using the providers' score list and hashData)
+func PickProviders(ctx sdk.Context, scores []*PairingScore, groupCount int, hashData []byte, chosenProvidersIdx map[int]bool) (returnedProviders []epochstoragetypes.StakeEntry) {
+	if len(scores) == 0 {
+		return returnedProviders
+	}
+
+	var scoreSum uint64
+	for idx, providerScore := range scores {
+		if chosenProvidersIdx[idx] {
+			// skip index of providers already selected
+			continue
+		}
+		scoreSum += providerScore.Score
+	}
+	if scoreSum == 0 {
+		err := fmt.Errorf("score sum is zero. Cannot pick providers for pairing")
+		panic(err)
+	}
+
+	if groupCount >= len(scores)-len(chosenProvidersIdx) {
+		providers := []epochstoragetypes.StakeEntry{}
+		for _, score := range scores {
+			providers = append(providers, *score.Provider)
+		}
+		return providers
+	}
+
+	for it := 0; it < groupCount; it++ {
+		hash := tendermintcrypto.Sha256(hashData) // TODO: we use cheaper algo for speed
```

Comment:
bug: you are using the same hashData for different groups resulting in the same number, this is not pseudo random and not fail, you need a differentiator in hashData for each slot
in the previous algo we did everything in one loop and the loop iterator was the differentiator but since your differentiator is starting form 0 all the time it's a mistake
i want a test that verifies pairing is uniformly sitributed pseudo random by scores

Reply:
done. see https://github.com/lavanet/lava/commit/0e1e5bc0039df5af58310188c100a4cdb2383d05